### PR TITLE
Generate crosswalk and import cap pdf

### PR DESCRIPTION
**Pull request by jtmst**
# Import Harvard Case Law Access Project (CAP) PDFs to CourtListener

## Issues Addressed:
- [Import CAP PDFs to CL](https://github.com/orgs/freelawproject/projects/54/views/1?pane=issue&itemId=72804379)
- [Generate crosswalk between CAP and CL](https://github.com/orgs/freelawproject/projects/54/views/1?pane=issue&itemId=72804762)

## Changes Implemented:
1. Created a management command to generate crosswalk files between CAP and CL data.
2. Developed a management command to import CAP PDFs to CL using the generated crosswalk.
3. Implemented S3 storage integration for PDF storage (configurable for local storage in development).
4. Added error handling and logging for better debugging and monitoring.
5. Added settings for CAP R2 env variables
6. Added tests for both commands

### Prerequisites:
- Ensure you have the necessary environment variables set for R2 and S3 access.
- For local testing, configure the storage to use local file system instead of S3.

### Steps:

1. Generate the crosswalk:
   ```
    docker exec -it cl-django python /opt/courtlistener/manage.py generate_cap_crosswalk
   ```
   This command will create crosswalk files in `cl/search/crosswalks/`.

2. Import CAP PDFs:
   ```
    docker exec -it cl-django python /opt/courtlistener/manage.py import_harvard_pdfs
   ```
   This command will use the generated crosswalk to fetch and store PDFs.

3. To test with specific parameters:
   ```
    docker exec -it cl-django python /opt/courtlistener/manage.py generate_cap_crosswalk --reporter "A.2d" --volume 100

   ```

### Verification:
- Check the `cl/search/crosswalks/` directory for generated crosswalk files.
- Verify that PDFs are stored either in S3 or local storage (based on configuration).
- Examine the logs for any errors or warnings during the process.

### CAP Crosswalk File (Sample Data):
Crosswalk File Sample

### Imported PDFs (Local Storage):
Imported PDFs

### CODE REVIEW

**Review by FRodriguez18 — COMMENTED** [review]
Looks good! Didn't approve it because I think John proposed some changes, but this looks good to me 👍

File: cl/search/tests/test_generate_cap_crosswalk.py
Comment (jtmst) [reply cl/search/tests/test_generate_cap_crosswalk.py]: It was not, missed that on cleanup, fixed!

**Comment by mlissner** [comment]
@flooie I just put this on your backlog to prioritize. The background is that we want to have the Harvard PDFs imported into CL and we want to do so regularly. Grab me when you have a sec, and I can give the details before you review or assign to somebody else to review.

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (quevon24) [inline cl/search/management/commands/generate_cap_crosswalk.py:171]: It would be great if you could check the citation with get_citations() from eyecite and use the data from the returned object to filter the OpinionCluster objects, here is an example: https://github.com/freelawproject/courtlistener/blob/main/cl/corpus_importer/management/commands/harvard_opinions.py#L636

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (quevon24) [inline cl/search/management/commands/generate_cap_crosswalk.py:153]: list_objects_v2 has a limit of 1000 objects, is there a possibility to have more than 1000?

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (quevon24) [inline cl/search/management/commands/generate_cap_crosswalk.py:91]: A cool way to know that your command did not freeze 👌

File: cl/search/management/commands/import_harvard_pdfs.py
Comment (quevon24) [inline cl/search/management/commands/import_harvard_pdfs.py:21]: Maybe this could be in cl/lib/storage.py, what do you think @flooie?

File: cl/search/tests/test_generate_cap_crosswalk.py
Comment (quevon24) [inline cl/search/tests/test_generate_cap_crosswalk.py:85]: You could also delete the file using tearDownClass method

File: cl/search/tests/test_import_harvard_pdfs.py
Comment (quevon24) [inline cl/search/tests/test_import_harvard_pdfs.py:45]: Here you could optionally have used DocketFactory and OpinionClusterFactory. Although as it is it is also valid.

**Review by quevon24 — COMMENTED** [review]
In general, it is necessary to add typing to the functions in addition to updating the docstrings of the functions so that they are in accordance with the format used in courtlistener.

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py:153]: From my understanding reporter volumes are limited at 999 before they are rolled into a new series, so this should not be an issue

Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py:153]: On second thought, I cant say that with certainty there is never 1000+.   Good suggestion, will add pagination to be safe

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py:171]: I've added this, but I'll note this could affect performance when run against all the data.  Run speed doesn't seem to be a huge concern at the moment, but something to note for later if the run time gets to get out of control in testing.

Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py:171]: Agreed that this will help accuracy though, thanks for the suggestion

File: cl/search/management/commands/import_harvard_pdfs.py
Comment (flooie) [reply cl/search/management/commands/import_harvard_pdfs.py:21]: Yeah - I think we should move anything storage related together

Comment (jtmst) [reply cl/search/management/commands/import_harvard_pdfs.py]: Makes sense, moved that to storage.py

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (quevon24) [inline cl/search/management/commands/generate_cap_crosswalk.py:209]: The docstrings for this function are missing

**Review by quevon24 — COMMENTED** [review]
Everything looks good, you've addressed the comments I left, one function is still missing the docstrings, and a comment I left regarding the find_matching_case() function is still pending.

After solving what I mentioned before, I think it will be ready.

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py]: In this case I suppose that would be the result, but I was under the impression that when this occurs the citation would be differentiated by page number as "100 A.2d 36a" or "100 A.2d 36b".  Is this not the case? 

If they were the exact same citation starting on the same page then there would be multiple matches and a logged warning with the citations. 

If this is a more common occurrence than I thought then I'll need to find a way to be more specific, how would CL differentiate between the two in this case, by docket number?

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py]: added

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (flooie) [inline cl/search/management/commands/generate_cap_crosswalk.py:247]: I am concerned about this part of the code.  

Why would we not use the Harvard CAP identifier to match up opinions.  We kept the Harvard XML file path field in our db.  Wouldn't it be easier (assuming you have that matched to the PDFs)  to use the specific identifier to the identifier found in the harvard file path?

Comment (mlissner) [reply cl/search/management/commands/generate_cap_crosswalk.py:247]: We also talked in our meeting today about how the same citation could map to two different scans. Imagine if page 31 of some book has two cases on it. At the top of the page, it has `Mike v. Jessie`, which is half a page long. Around the middle of the page, `Frank v. Howard` begins, but then goes on for another five pages.

In that case, both cases will have the same citation, but the scan for the first case will be one page long, while the scan for the second will be six pages long.

Comment (jtmst) [reply cl/search/management/commands/generate_cap_crosswalk.py]: I've updated this to instead query the filepath field for a matching path constructed from the case data and cap identifier.  Using this approach we shouldnt need to worry about cases with matching citations/

**Review by quevon24 — APPROVED** [review]
Everything looks very good, using the cap id greatly simplified the match, I have no more comments

**Comment by mlissner** [comment]
I think we're just waiting for @flooie's final review here, but he's out sick today. Hopefully we'll get this merged shortly. Sorry for the delay.

**Review by flooie — APPROVED** [review]
This looks good to me-  Thanks for taking the time to rewrite the matching code.

File: cl/settings/third_party/r2.py
Comment (mlissner) [inline cl/settings/third_party/r2.py:1]: I think I'd be confused if i looked in r2.py and only found CAP settings. Could we move these settings into misc.py instead?

**Review by mlissner — CHANGES_REQUESTED** [review]
Sorry to chime in a bit late. I came in here to see if there were any DB migrations before merging, but I noticed a couple of last minute things I think we should tweak. 

@flooie or @quevon24, do you think we could make these changes so Josh can consider his part of this done?

File: cl/search/management/commands/generate_cap_crosswalk.py
Comment (quevon24) [reply cl/search/management/commands/generate_cap_crosswalk.py]: i made the --output-dir argument required and removed the default directory

File: cl/settings/third_party/r2.py
Comment (quevon24) [reply cl/settings/third_party/r2.py:1]: I moved the settings to misc.py

File: cl/search/crosswalks/example_crosswalk.json
Comment (quevon24) [reply cl/search/crosswalks/example_crosswalk.json:1]: I couldn't find any use for the file other than documentation purposes, so I deleted it and added it as a comment in the generate_cap_crosswalk command

Comment (flooie) [reply cl/search/crosswalks/example_crosswalk.json:1]: Thanks @quevon24 and @mlissner

**Comment by flooie** [comment]
@mlissner Thanks for your suggestions.  The failing test passed on a re-run and appears to be related to an error in the meeting I missed.

**Review by mlissner — APPROVED** [review]

**Comment by quevon24** [comment]
> get



> Woot! We're merged. @quevon24, do you want to file an issue to get this deployed and work with Ramiro on landing the changes?

Yes, I'll get to it, I think we would only need the values ​​of the environment variables necessary for the command, or are they already set?

**Comment by mlissner** [comment]
They're not, I don't think. I'll get them to Ramiro.

**Pull request merged**
